File: ts/project/.sdk/tm/java/utility/Json.java
package JAVAPACKAGE.utility;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Minimal zero-dependency JSON parser for the ProjectName SDK runtime.
 * Produces the same shapes the rest of the runtime consumes:
 * LinkedHashMap&lt;String,Object&gt;, ArrayList&lt;Object&gt;, String,
 * Long (integer-valued numbers), Double, Boolean and null.
 *
 * Serialization is handled by the vendored Struct.jsonify; this class only
 * parses. {@link #parse} throws IllegalArgumentException on malformed input;
 * {@link #parseOrNull} returns null instead (used by the HTTP fetcher for
 * non-JSON bodies).
 */
public final class Json {

  private Json() {}

  public static Object parse(String src) {
    if (src == null) {
      throw new IllegalArgumentException("json: null input");
    }
    Parser p = new Parser(src);
    p.ws();
    Object out = p.value();
    p.ws();
    if (!p.done()) {
      throw new IllegalArgumentException("json: trailing content at " + p.pos);
    }
    return out;
  }

  public static Object parseOrNull(String src) {
    try {
      return parse(src);
    }
    catch (RuntimeException e) {
      return null;
    }
  }

  private static final class Parser {
    final String s;
    int pos = 0;

    Parser(String s) {
      this.s = s;
    }

    boolean done() {
      return pos >= s.length();
    }

    void ws() {
      while (pos < s.length()) {
        char c = s.charAt(pos);
        if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
          pos++;
        }
        else {
          break;
        }
      }
    }

    char peek() {
      if (pos >= s.length()) {
        throw new IllegalArgumentException("json: unexpected end of input");
      }
      return s.charAt(pos);
    }

    void expect(char c) {
      if (peek() != c) {
        throw new IllegalArgumentException(
            "json: expected '" + c + "' at " + pos + ", got '" + peek() + "'");
      }
      pos++;
    }

    Object value() {
      char c = peek();
      switch (c) {
        case '{':
          return object();
        case '[':
          return array();
        case '"':
          return string();
        case 't':
          literal("true");
          return Boolean.TRUE;
        case 'f':
          literal("false");
          return Boolean.FALSE;
        case 'n':
          literal("null");
          return null;
        default:
          if (c == '-' || (c >= '0' && c <= '9')) {
            return number();
          }
          throw new IllegalArgumentException("json: unexpected '" + c + "' at " + pos);
      }
    }

    void literal(String lit) {
      if (!s.startsWith(lit, pos)) {
        throw new IllegalArgumentException("json: invalid literal at " + pos);
      }
      pos += lit.length();
    }

    Map<String, Object> object() {
      expect('{');
      Map<String, Object> out = new LinkedHashMap<>();
      ws();
      if (peek() == '}') {
        pos++;
        return out;
      }
      while (true) {
        ws();
        String key = string();
        ws();
        expect(':');
        ws();
        out.put(key, value());
        ws();
        char c = peek();
        if (c == ',') {
          pos++;
          continue;
        }
        if (c == '}') {
          pos++;
          return out;
        }
        throw new IllegalArgumentException("json: expected ',' or '}' at " + pos);
      }
    }

    List<Object> array() {
      expect('[');
      List<Object> out = new ArrayList<>();
      ws();
      if (peek() == ']') {
        pos++;
        return out;
      }
      while (true) {
        ws();
        out.add(value());
        ws();
        char c = peek();
        if (c == ',') {
          pos++;
          continue;
        }
        if (c == ']') {
          pos++;
          return out;
        }
        throw new IllegalArgumentException("json: expected ',' or ']' at " + pos);
      }
    }

    String string() {
      expect('"');
      StringBuilder b = new StringBuilder();
      while (true) {
        if (pos >= s.length()) {
          throw new IllegalArgumentException("json: unterminated string");
        }
        char c = s.charAt(pos++);
        if (c == '"') {
          return b.toString();
        }
        if (c == '\\') {
          if (pos >= s.length()) {
            throw new IllegalArgumentException("json: unterminated escape");
          }
          char e = s.charAt(pos++);
          switch (e) {
            case '"':
              b.append('"');
              break;
            case '\\':
              b.append('\\');
              break;
            case '/':
              b.append('/');
              break;
            case 'b':
              b.append('\b');
              break;
            case 'f':
              b.append('\f');
              break;
            case 'n':
              b.append('\n');
              break;
            case 'r':
              b.append('\r');
              break;
            case 't':
              b.append('\t');
              break;
            case 'u':
              if (pos + 4 > s.length()) {
                throw new IllegalArgumentException("json: bad unicode escape");
              }
              b.append((char) Integer.parseInt(s.substring(pos, pos + 4), 16));
              pos += 4;
              break;
            default:
              throw new IllegalArgumentException("json: bad escape '\\" + e + "'");
          }
        }
        else {
          b.append(c);
        }
      }
    }

    Object number() {
      int start = pos;
      if (peek() == '-') {
        pos++;
      }
      while (pos < s.length() && Character.isDigit(s.charAt(pos))) {
        pos++;
      }
      boolean integral = true;
      if (pos < s.length() && s.charAt(pos) == '.') {
        integral = false;
        pos++;
        while (pos < s.length() && Character.isDigit(s.charAt(pos))) {
          pos++;
        }
      }
      if (pos < s.length() && (s.charAt(pos) == 'e' || s.charAt(pos) == 'E')) {
        integral = false;
        pos++;
        if (pos < s.length() && (s.charAt(pos) == '+' || s.charAt(pos) == '-')) {
          pos++;
        }
        while (pos < s.length() && Character.isDigit(s.charAt(pos))) {
          pos++;
        }
      }
      String num = s.substring(start, pos);
      try {
        if (integral) {
          return Long.parseLong(num);
        }
        return Double.parseDouble(num);
      }
      catch (NumberFormatException e) {
        return Double.parseDouble(num);
      }
    }
  }
}

File: ts/project/.sdk/tm/kotlin/utility/Json.kt
package KOTLINPACKAGE.utility

/**
 * Minimal zero-dependency JSON parser for the ProjectName SDK runtime.
 * Produces the same shapes the rest of the runtime consumes:
 * LinkedHashMap<String, Any?>, ArrayList<Any?>, String, Long (integer-valued
 * numbers), Double, Boolean and null.
 *
 * Serialization is handled by the vendored Struct.jsonify; this class only
 * parses. [parse] throws IllegalArgumentException on malformed input;
 * [parseOrNull] returns null instead (used by the HTTP fetcher for non-JSON
 * bodies).
 */
object Json {

  fun parse(src: String?): Any? {
    if (src == null) {
      throw IllegalArgumentException("json: null input")
    }
    val p = Parser(src)
    p.ws()
    val out = p.value()
    p.ws()
    if (!p.done()) {
      throw IllegalArgumentException("json: trailing content at " + p.pos)
    }
    return out
  }

  fun parseOrNull(src: String?): Any? {
    return try {
      parse(src)
    } catch (e: RuntimeException) {
      null
    }
  }

  private class Parser(val s: String) {
    var pos: Int = 0

    fun done(): Boolean {
      return pos >= s.length
    }

    fun ws() {
      while (pos < s.length) {
        val c = s[pos]
        if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
          pos++
        } else {
          break
        }
      }
    }

    fun peek(): Char {
      if (pos >= s.length) {
        throw IllegalArgumentException("json: unexpected end of input")
      }
      return s[pos]
    }

    fun expect(c: Char) {
      if (peek() != c) {
        throw IllegalArgumentException("json: expected '$c' at $pos, got '${peek()}'")
      }
      pos++
    }

    fun value(): Any? {
      val c = peek()
      return when (c) {
        '{' -> obj()
        '[' -> array()
        '"' -> string()
        't' -> {
          literal("true")
          true
        }
        'f' -> {
          literal("false")
          false
        }
        'n' -> {
          literal("null")
          null
        }
        else -> {
          if (c == '-' || (c in '0'..'9')) {
            number()
          } else {
            throw IllegalArgumentException("json: unexpected '$c' at $pos")
          }
        }
      }
    }

    fun literal(lit: String) {
      if (!s.startsWith(lit, pos)) {
        throw IllegalArgumentException("json: invalid literal at $pos")
      }
      pos += lit.length
    }

    fun obj(): MutableMap<String, Any?> {
      expect('{')
      val out = linkedMapOf<String, Any?>()
      ws()
      if (peek() == '}') {
        pos++
        return out
      }
      while (true) {
        ws()
        val key = string()
        ws()
        expect(':')
        ws()
        out[key] = value()
        ws()
        val c = peek()
        if (c == ',') {
          pos++
          continue
        }
        if (c == '}') {
          pos++
          return out
        }
        throw IllegalArgumentException("json: expected ',' or '}' at $pos")
      }
    }

    fun array(): MutableList<Any?> {
      expect('[')
      val out = mutableListOf<Any?>()
      ws()
      if (peek() == ']') {
        pos++
        return out
      }
      while (true) {
        ws()
        out.add(value())
        ws()
        val c = peek()
        if (c == ',') {
          pos++
          continue
        }
        if (c == ']') {
          pos++
          return out
        }
        throw IllegalArgumentException("json: expected ',' or ']' at $pos")
      }
    }

    fun string(): String {
      expect('"')
      val b = StringBuilder()
      while (true) {
        if (pos >= s.length) {
          throw IllegalArgumentException("json: unterminated string")
        }
        val c = s[pos++]
        if (c == '"') {
          return b.toString()
        }
        if (c == '\\') {
          if (pos >= s.length) {
            throw IllegalArgumentException("json: unterminated escape")
          }
          val e = s[pos++]
          when (e) {
            '"' -> b.append('"')
            '\\' -> b.append('\\')
            '/' -> b.append('/')
            'b' -> b.append('\b')
            'f' -> b.append('\u000C')
            'n' -> b.append('\n')
            'r' -> b.append('\r')
            't' -> b.append('\t')
            'u' -> {
              if (pos + 4 > s.length) {
                throw IllegalArgumentException("json: bad unicode escape")
              }
              b.append(s.substring(pos, pos + 4).toInt(16).toChar())
              pos += 4
            }
            else -> throw IllegalArgumentException("json: bad escape '\\$e'")
          }
        } else {
          b.append(c)
        }
      }
    }

    fun number(): Any {
      val start = pos
      if (peek() == '-') {
        pos++
      }
      while (pos < s.length && s[pos].isDigit()) {
        pos++
      }
      var integral = true
      if (pos < s.length && s[pos] == '.') {
        integral = false
        pos++
        while (pos < s.length && s[pos].isDigit()) {
          pos++
        }
      }
      if (pos < s.length && (s[pos] == 'e' || s[pos] == 'E')) {
        integral = false
        pos++
        if (pos < s.length && (s[pos] == '+' || s[pos] == '-')) {
          pos++
        }
        while (pos < s.length && s[pos].isDigit()) {
          pos++
        }
      }
      val num = s.substring(start, pos)
      return try {
        if (integral) {
          num.toLong()
        } else {
          num.toDouble()
        }
      } catch (e: NumberFormatException) {
        num.toDouble()
      }
    }
  }
}
